File: 11b.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

my @monkeys;
my $product = 1;

while (<>) {
    if (/^Monkey [0-9]+:/) {
        push @monkeys, {};

    } elsif (/Starting items: ([0-9, ]*)/) {
        $monkeys[-1]{items} = [split /, /, $1];

    } elsif (/Operation: new = (.*)/) {
        $monkeys[-1]{operation} = $1;

    } elsif (/Test: divisible by ([0-9]+)/) {
        $monkeys[-1]{test} = $1;
        $product *= $1;

    } elsif (/If (true|false): throw to monkey ([0-9]+)/) {
        $monkeys[-1]{$1} = $2;

    } elsif (/^.+$/) {
        die "Can't parse $_.\n";
    }
}

for my $round (1 .. 10_000) {
    for my $monkey (@monkeys) {
        while (@{ $monkey->{items} }) {
            ++$monkey->{active};
            my $item = shift @{ $monkey->{items} };
            my $expression = $monkey->{operation};
            $expression =~ s/old/$item/g;
            my @tokens = $expression =~ /([0-9]+) ([+*]) ([0-9]+)/;
            my $level = $tokens[1] eq '+' ? $tokens[0] + $tokens[2]
                                          : $tokens[0] * $tokens[2];

            my $to = $monkey->{ (0 == $level % $monkey->{test}) ? 'true'
                                                                : 'false' };
            push @{ $monkeys[$to]{items} }, $level % $product;
        }
    }
}

my @business = (0, 0);
for my $monkey (@monkeys) {
    if ($monkey->{active} >= $business[0]) {
        splice @business,
            ($monkey->{active} >= $business[1])
                ? 2
                : 1,
            0,
            $monkey->{active};
        shift @business;
    }
}
say $business[0] * $business[1];


__DATA__
Monkey 0:
  Starting items: 79, 98
  Operation: new = old * 19
  Test: divisible by 23
    If true: throw to monkey 2
    If false: throw to monkey 3

Monkey 1:
  Starting items: 54, 65, 75, 74
  Operation: new = old + 6
  Test: divisible by 19
    If true: throw to monkey 2
    If false: throw to monkey 0

Monkey 2:
  Starting items: 79, 60, 97
  Operation: new = old * old
  Test: divisible by 13
    If true: throw to monkey 1
    If false: throw to monkey 3

Monkey 3:
  Starting items: 74
  Operation: new = old + 3
  Test: divisible by 17
    If true: throw to monkey 0
    If false: throw to monkey 1

File: 11b.py
#! /usr/bin/python3
import re
import sys

monkeys = []
product = 1
with open(sys.argv[1], 'r') as fh:
    for line in fh:
        if re.match(r'Monkey [0-9]+:', line):
            monkeys += [{'active':0}]
        elif re.search(r'Starting items', line):
            s = re.search(r'([0-9][^\n]+)', line)
            monkeys[-1]['items'] = s.group(1).split(', ')
        elif re.search(r'Operation:', line):
            s = re.search(r'new = (.*)', line)
            monkeys[-1]['operation'] = s.group(1)
        elif re.search(r'Test:', line):
            s = re.search(r'by ([0-9]+)', line)
            t = int(s.group(1))
            monkeys[-1]['test'] = t
            product *= t
        elif re.search(r'throw to monkey', line):
            s = re.search(r'(true|false): throw to monkey ([0-9]+)', line)
            monkeys[-1][s.group(1)] = s.group(2)
        elif re.search(r'.', line):
            raise Exception(f'Can\'t parse {line}')

for round in range(1, 10001):
    for monkey in monkeys:
        while monkey['items']:
            monkey['active'] += 1
            item = monkey['items'].pop(0)
            expression = monkey['operation']
            v1, op, v2 = expression.replace('old', item).split()
            level = int(v1) + int(v2) if '+' == op else int(v1) * int(v2)
            to = monkey['true' if 0 == level % monkey['test'] else 'false']
            monkeys[int(to)]['items'] += [str(level % product)]

business = [0, 0]
for monkey in monkeys:
    if monkey['active'] >= business[0]:
        business += [monkey['active']]
        business.sort()
        business.pop(0)
print(business[0] * business[1])
